services/venice-controller/src/main/java/com/linkedin/venice/controller/UnusedValueSchemaCleanupService.java

```diff
@@ -0,0 +1,127 @@
+  public final int minSchemaCountToKeep;
+
+  UnusedValueSchemaCleanupService(
+      VeniceControllerMultiClusterConfig multiClusterConfig,
+      VeniceHelixAdmin admin,
+      VeniceParentHelixAdmin parentHelixAdmin) {
+    this.multiClusterConfig = multiClusterConfig;
+    this.admin = admin;
+    this.scheduleIntervalMinutes = multiClusterConfig.getUnusedSchemaCleanupIntervalMinutes();
+    this.minSchemaCountToKeep = multiClusterConfig.getMinSchemaCountToKeep();
+    this.veniceParentHelixAdmin = parentHelixAdmin;
+  }
+
+  private Runnable getRunnableForSchemaCleanup() {
+    return () -> {
+      if (stop) {
+        return;
+      }
+      for (String clusterName: multiClusterConfig.getClusters()) {
+        boolean cleanupEnabled =
+            multiClusterConfig.getControllerConfig(clusterName).isUnusedValueSchemaCleanupServiceEnabled();
+        if (!cleanupEnabled) {
+          continue;
+        }
+        // Get all stores for current cluster
+        List<Store> stores = admin.getAllStores(clusterName);
+        for (Store store: stores) {
+          String storeName = store.getName();
+          // Remove schema only for batch stores
+          if (isHybridStore(store) || VeniceSystemStoreUtils.isSystemStore(storeName)) {
+            continue;
+          }
+          Collection<SchemaEntry> allSchemas = veniceParentHelixAdmin.getValueSchemas(clusterName, storeName);
+
+          if (allSchemas.size() < minSchemaCountToKeep) {
+            continue;
+          }
+
+          Set<Integer> usedSchemaSet = veniceParentHelixAdmin.getInUseValueSchemaIds(clusterName, storeName);
+
+          if (usedSchemaSet.isEmpty()) {
+            continue;
+          }
+          int minSchemaIdInUse = Collections.min(usedSchemaSet);
+          Set<Integer> schemasToDelete = new HashSet<>();
+
+          // assumes `getValueSchemas` returns ascending schema ids so that the older schemas are deleted first
+          for (SchemaEntry schemaEntry: allSchemas) {
+            int schemaId = schemaEntry.getId();
+            // skip latest value schema or super-set schema id
+            if (schemaId == store.getLatestSuperSetValueSchemaId() || admin.getHelixVeniceClusterResources(clusterName)
+                .getSchemaRepository()
+                .getSupersetOrLatestValueSchema(storeName)
+                .getId() == schemaId) {
+              continue;
+            }
+
+            // delete only if its not used and less than minimum of used schema id
+            if (!usedSchemaSet.contains(schemaId) && schemaId < minSchemaIdInUse) {
+              schemasToDelete.add(schemaId);
+              // maintain minimum of SCHEMA_COUNT_THRESHOLD schemas in repo
+              if (schemasToDelete.size() > allSchemas.size() - minSchemaCountToKeep) {
+                break;
+              }
+            }
+          }
+          // delete from child colos
+          admin.deleteValueSchemas(clusterName, store.getName(), schemasToDelete);
```

Comment:
This ordering is important, in this case since the service is running in parent controller we should actually call `veniceParentHelixAdmin.deleteValueSchemas` before ` admin.deleteValueSchemas`. This way if one of the colo is not reachable or throwing exceptions we get to retry the deletion on the next run because we haven't deleted the unused schemas in parent yet. We should add an unit test for this scenario (where one colo is temporarily unreachable).

Reply:
Good point!  I will add a test case for that.